# 앙상블

## (1) Voting

* 위스콘신 유방암 dataset 을 이용한 실습

In [4]:

import pandas as pd

from sklearn.ensemble import VotingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# dataset 로드
from sklearn.datasets import load_breast_cancer

In [5]:
# 데이터 프레임에 담아주기
cancer = load_breast_cancer()

data_df = pd.DataFrame(cancer.data, columns=cancer.feature_names)
data_df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst radius  worst texture  worst perimeter  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   worst area  worst smoothness  worst compactness  worst concavity  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   worst concave points  worst symmetry  worst fractal dimension  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [6]:
## 개별 모델은 로지스틱 회귀와 KNN 임
Ir_clf = LogisticRegression()
Knn_clf = KNeighborsClassifier(n_neighbors=8)

In [7]:
# 개별 모델을 소프트 보딩 기반의 앙상블 모델로 구현한 분류기
vo_clf = VotingClassifier(estimators= [('LR', Ir_clf),('KNN', Knn_clf)], voting='soft')

X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, test_size= 0.2, random_state=156)


# voting 분류 학습/예측/평가
vo_clf.fit(X_train, y_train)
pred = vo_clf.predict(X_test)
print("Voting 분류기 정확도 :{0: 4f}".format(accuracy_score(y_test, pred)))


# 개별 모델의 학습/ 예측/ 평가
classifiers = [Ir_clf, Knn_clf]
for classfier in classifiers :
    classfier.fit(X_train, y_train)
    pred = classfier.predict(X_test)
    class_name = classfier.__class__.__name__
    print("{0} 정확도: {1: .4f}".format(class_name, accuracy_score(y_test, pred)))

Voting 분류기 정확도 : 0.947368
LogisticRegression 정확도:  0.9386
KNeighborsClassifier 정확도:  0.9386


C:\Users\LG\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\LG\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/m

Voting 분류기 정확도 : 0.947368
LogisticRegression 정확도:  0.9386
KNeighborsClassifier 정확도:  0.9386
c:\Users\Jin\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\Users\Jin\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(

## (2) Bagging : 랜덤 포레스트


* 랜덤 포레스트 : Bagging(배깅)의 대표적인 알고리즘

sklearn.ensemble.RandomForestClassifier

a. import 방법 : from sklearn.ensemble import RandomForestClassifier

b. 공식홈 : https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html?highlight=randomforest#sklearn.ensemble.RandomForestClassifier

c. parameter 
> n_estimators=100, \
> criterion='gini', \
> max_depth=None, \
> min_samples_split=2, \
> min_samples_leaf=1, \
> min_weight_fraction_leaf=0.0,\ 
> max_features='sqrt', \
> max_leaf_nodes=None, \
> min_impurity_decrease=0.0,\
> bootstrap=True, \
> oob_score=False, \
> n_jobs=None, \
> random_state=None, \
> verbose=0, \
> warm_start=False, \
> class_weight=None, \
> ccp_alpha=0.0, \
> max_samples=None
-----------------------------------------------------